std-bits/table/src/main/java/org/enso/table/data/index/CrossTabIndex.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    // Fill numberings
+    xKeyNumberer = new ObjectNumberer<>(xSubKeys);
+    yKeyNumberer = new ObjectNumberer<>(ySubKeys);
+
+    // Create grid of cells, mapping x and y key indices to combined keys.
+    grid = new UnorderedMultiValueKey[numXKeys()][numYKeys()];
+
+    // For each combined key, use the two subkeys to determine row+col
+    // coordinates, and put the key at those coordinates.
+    for (int i = 0; i < combinedIndex.size(); ++i) {
+      UnorderedMultiValueKey combinedKey = combinedKeys.get(i);
+      int xCoordinate = getXCoordinate(xSubKeys.get(i));
+      int yCoordinate = getYCoordinate(ySubKeys.get(i));
+
+      // The pair (xCoordinate, yCoordinate) must be unique so this
+      // check is not really necessary.
+      if (grid[xCoordinate][yCoordinate] != null) {
+        throw new IllegalStateException("Internal error: makeCrossTabTable coordinate conflict");
+      }
+
+      grid[xCoordinate][yCoordinate] = combinedKey;
+    }
+  }
+
+  public List<Integer> get(UnorderedMultiValueKey xKey, UnorderedMultiValueKey yKey) {
+    return combinedIndex.get(grid[getXCoordinate(xKey)][getYCoordinate(yKey)]);
+  }
+
+  public Set<UnorderedMultiValueKey> getXKeys() {
+    return xKeyNumberer.getObjects();
+  }
+
+  public Set<UnorderedMultiValueKey> getYKeys() {
+    return yKeyNumberer.getObjects();
+  }
+
+  public int getXCoordinate(UnorderedMultiValueKey key) {
+    return xKeyNumberer.getNumber(key);
+  }
+
+  public int getYCoordinate(UnorderedMultiValueKey key) {
+    return yKeyNumberer.getNumber(key);
+  }
+
+  public int numXKeys() {
+    return xKeyNumberer.size();
+  }
+
+  public int numYKeys() {
+    return yKeyNumberer.size();
+  }
+
+  public Table makeCrossTabTable(Aggregator[] aggregates, String[] aggregateNames) {
+    Context context = Context.getCurrent();
+    NameDeduplicator outputTableNameDeduplicator = new NameDeduplicator();
+
+    final int columnCount = yColumns.length + numXKeys() * aggregates.length;
+    if (columnCount > MAXIMUM_CROSS_TAB_COLUMN_COUNT) {
+      throw new TooManyColumnsException(
+          "The cross_tab contained too many columns. Maximum allowed is "
+              + MAXIMUM_CROSS_TAB_COLUMN_COUNT
+              + " but was "
+              + columnCount
+              + ".",
+          columnCount,
+          MAXIMUM_CROSS_TAB_COLUMN_COUNT);
+    }
+
+    // Create the storage
+    Builder[] storage = new Builder[columnCount];
+    for (int i = 0; i < yColumns.length; i++) {
+      storage[i] = Builder.getForType(yColumns[i].getStorage().getType(), numYKeys());
+      context.safepoint();
+    }
+
+    for (int i = 0; i < numXKeys(); i++) {
+      int offset = yColumns.length + i * aggregates.length;
+      for (int j = 0; j < aggregates.length; j++) {
+        storage[offset + j] = Builder.getForType(aggregates[j].getType(), numYKeys());
+        context.safepoint();
+      }
+    }
+
+    // Fill the columns.
+    var emptyList = new ArrayList<Integer>();
```

Comment:
```suggestion
    var emptyList = List.of<Integer>();
```

That way it will be guaranteed immutable.

Reply:
Actually not sure if this variable is needed, what about just doing:
```
rowIds = List.of();
```
later on?

Reply:
Done.